sync/src/synchronizer/mod.rs

```diff
@@ -357,33 +418,112 @@ impl Synchronizer {
+                // best known < tip and in IBD state, these node can be disconnect
+                list.extend(
+                    self.shared
+                        .state()
+                        .peers()
+                        .get_best_known_less_than_tip(tip),
+                )
+            };
+            list
+        };
+
+        for peer in disconnect_list.iter() {
+            if self
+                .peers()
+                .get_flag(*peer)
+                .map(|flag| flag.is_whitelist || flag.is_protect)
+                .unwrap_or(false)
+            {
+                continue;
+            }
+            if let Err(err) = nc.disconnect(*peer, "sync disconnect") {
+                debug!("synchronizer disconnect error: {:?}", err);
+            }
+        }
+
+        if !self.can_fetch_block.load(Ordering::Acquire) {
+            return;
+        }
+
         let peers: Vec<PeerIndex> = {
-            self.peers()
+            let state = &self
+                .shared
+                .state()
+                .read_inflight_blocks()
+                .download_schedulers;
+            let mut peers: Vec<PeerIndex> = self
+                .peers()
                 .state
                 .read()
                 .iter()
-                .filter(|(_, state)| match ibd {
-                    IBDState::In => {
-                        state.peer_flags.is_outbound
-                            || state.peer_flags.is_whitelist
-                            || state.peer_flags.is_protect
+                .filter(|(id, state)| {
+                    if disconnect_list.contains(id) {
+                        return false;
+                    };
+                    match ibd {
+                        IBDState::In => {
+                            state.peer_flags.is_outbound
+                                || state.peer_flags.is_whitelist
+                                || state.peer_flags.is_protect
+                        }
+                        IBDState::Out => state.sync_started,
                     }
-                    IBDState::Out => state.sync_started,
                 })
                 .map(|(peer_id, _)| peer_id)
                 .cloned()
-                .collect()
+                .collect();
+            peers.sort_by_key(|id| {
+                state
+                    .get(id)
+                    .map(|d| d.task_count())
+                    .unwrap_or(crate::MAX_BLOCKS_IN_TRANSIT_PER_PEER)
+            });
+            peers.reverse();
+            peers
         };
 
         trace!("poll find_blocks_to_fetch select peers");
-        {
-            self.shared().state().write_inflight_blocks().prune();
-        }
-        for peer in peers {
-            if let Some(fetch) = self.get_blocks_to_fetch(peer, ibd) {
-                if !fetch.is_empty() {
-                    self.send_getblocks(fetch, nc, peer);
+        // fetch use a lot of cpu time, especially in ibd state
```

Comment:
This comment only talks about a reason, it lacks of a "so" clause to say how we fix this issue.

Reply:
so in ibd, I use a thread to fetch block